smac_planner/include/smac_planner/costmap_downsampler.hpp

```diff
@@ -0,0 +1,186 @@
+  }
+
+  /**
+   * @brief Activate the publisher of the downsampled costmap
+   */
+  void activatePublisher() {
+    _downsampled_costmap_pub->on_activate();
+  }
+
+  /**
+   * @brief Deactivate the publisher of the downsampled costmap
+   */
+  void deactivatePublisher() {
+    _downsampled_costmap_pub->on_deactivate();
+  }
+
+  /**
+   * @brief Downsample the given costmap by the downsampling factor, and publish the downsampled costmap
+   * @param downsampling_factor Multiplier for the costmap resolution
+   * @return A ptr to the downsampled costmap
+   */
+  nav2_costmap_2d::Costmap2D * downsample(const unsigned int & downsampling_factor)
+  {
+    unsigned int new_mx, new_my;
+    _downsampling_factor = downsampling_factor;
+    updateCostmapSize();
+
+    // Adjust costmap size if needed
+    if (_downsampled_costmap->getSizeInCellsX() != _downsampled_size_x ||
+      _downsampled_costmap->getSizeInCellsY() != _downsampled_size_y ||
+      _downsampled_costmap->getResolution() != _downsampled_resolution) {
+      resizeCostmap();
+    }
+
+    // Assign costs
+    for (int i = 0; i < _downsampled_size_x * _downsampled_size_y; ++i) {
+      new_mx = i % _downsampled_size_x;
+      new_my = i / _downsampled_size_x;
+      setCostOfCell(new_mx, new_my);
+    }
+
+    if (_node->count_subscribers(_topic_name) > 0) {
+      _downsampled_costmap_pub->publishCostmap();
+    }
+
+    return _downsampled_costmap.get();
+  }
+
+private:
+  /**
+   * Update the sizes X-Y of the costmap and its downsampled version
+   */
+  void updateCostmapSize()
+  {
+    _size_x = _costmap->getSizeInCellsX();
+    _size_y = _costmap->getSizeInCellsY();
+    _downsampled_size_x = ceil(static_cast<float>(_size_x) / _downsampling_factor);
+    _downsampled_size_y = ceil(static_cast<float>(_size_y) / _downsampling_factor);
+    _downsampled_resolution = _downsampling_factor * _costmap->getResolution();
+  }
+
+  /**
+   * Resize the downsampled costmap. Used in case the costmap changes and we need to update the downsampled version
+   */
+  void resizeCostmap()
+  {
+    _downsampled_costmap->resizeMap(
+      _downsampled_size_x,
+      _downsampled_size_y,
+      _downsampled_resolution,
+      _costmap->getOriginX(),
+      _costmap->getOriginY());
+  }
+
+  /**
+   * @brief Explore all subcells of the original costmap and assign the max cost to the new (downsampled) cell
+   * @param new_mx The X-coordinate of the cell in the new costmap
+   * @param new_my The Y-coordinate of the cell in the new costmap
+   */
+  void setCostOfCell(
+    const unsigned int & new_mx,
+    const unsigned int & new_my)
+  {
+    unsigned int mx, my;
+    unsigned char cost = 0;
+    unsigned int x_offset = new_mx * _downsampling_factor;
+    unsigned int y_offset = new_my * _downsampling_factor;
+
+    for(int j = 0; j < _downsampling_factor * _downsampling_factor; ++j) {
+      mx = std::min(x_offset + j % _downsampling_factor, _size_x - 1);
```

Comment:
Are these `min` checks required if you did the ceil correctly in the update step?

Reply:
I gave this some thought and normally I wouldn't be able to remove those checks. The downsampled costmap that I'm creating is _at least_ as big as the original costmap, but it can be a bit bigger depending on the size of the map and the sampling factor we chose. Having said that, this means that we can end up exploring cells that are out of bounds for the original costmap... But...

I ran a test in a 566x608 map of cells of 0.1 m, when we downsample to cells of 0.3m, we get a costmap of 189x203 cells. This means that when we are assigning costs to the index new_mx = 188 (corresponding to cells 189s) then we would explore cells 564, 565 and 566! Normally, when getting the cost of a cell with index 566 from the original costmap we would get an out of bounds error, but we are not. Not sure why this is happening, I see the cost associated to it is zero. It could be that it just explores uninitialized memory and interprets it as a byte (char) and it so happened to be zero?

Anyways, bottomline is we shouldn't remove it if we want to have well-defined behaviour. Removing it will likely go out-of-bounds for the charmap and have some undefined behavour near the edges of the map

Reply:
Yeah, so I just checked and there's no access safety when calling `_costmap->getCost(x, y)`. I tried checking for the cost at the cell 1000x1000 which is clearly out of bounds and got
`exploring costmap at 1000x1000 -> equal to 224`.

https://stackoverflow.com/questions/39188469/char-array-can-hold-more-than-expected

> As soon as you read or write from an array outside of its bounds, you invoke undefined behavior.
> 
> Whenever this happens, the program may do anything it wants. It is even allowed to play you a birthday song although it's not your birthday, or to transfer money from your bank account. Or it may crash, or delete files, or just pretend nothing bad happened.

Reply:
If you want to get rid of the `min`s, the only option would be to create the costmap being `at most` the same size as the original costmap, but likely smaller (opposite of what we do now). This way we know for sure we will never explore out-of-bound cells for the original costmap without the need of `min`s, at the price of having a slightly smaller downsampled map (whereas now we can a slightly larger one). Your call.

Reply:
This means simply changing `ceil`s to `floor`s in the `updateCostmapSize` method. I think it's minor not having those cells in the edges of the map, taking into account that normally the robot won't be going there anyway. If you think the optimization of not having the `min` calls outweights the loss of a few cells in the top and right edges of the map, then I'll send the commit. Otherwise I think we can merge as is.

Reply:
> The downsampled costmap that I'm creating is at least as big as the original costmap, but it can be a bit bigger depending on the size of the map and the sampling factor we chose

Shouldn't it be deterministic the size to be exactly what we ask the size to be? Why would it be hilariously larger? The costmap resize goes to this function https://github.com/ros-planning/navigation2/blob/aaa97902c1e1bb9a509c4ee0d88b880afb528f20/nav2_costmap_2d/src/costmap_2d.cpp#L75 which calls a delete and new allocation of the costmap of the exact size requested. 

> this means that we can end up exploring cells that are out of bounds for the original costmap

Yeah, that's really not good, its not clear to me why we'd be exploring `564, 565 and 566!` if the size is so much smaller, who's exploring this, exactly? The A* is only given the downsampled costmap which would be 3x smaller, it wouldn't know anything about the original costmap. It would be fine from the rounding if around the edges we had 1 cell (at most) extra because of rounding error but we should never be in a situation where we have some NxN costmap but we only use N-MxN-M of it and then we explore this non-deterministic space

Reply:
> Shouldn't it be deterministic the size to be exactly what we ask the size to be? Why would it be hilariously larger?

The size of the array is deterministic, what is undefined behaviour is accessing indeces of the array that are out-of-bounds. Check this https://stackoverflow.com/questions/1239938/accessing-an-array-out-of-bounds-gives-no-error-why. This explains why I can do `_costmap[1000 * 1000]` and get some value instead of an error.

> Yeah, that's really not good, its not clear to me why we'd be exploring 564, 565 and 566! if the size is so much smaller, who's exploring this, exactly?

Those cells are from the original costmap (the one that has more cells). We explore those in the SetCellCosts method from the downsampler, where we are checking every cell of the original map in a certain fashion to assign it to a corresponding downsampled cell. In general, when you downsample you risk going at most `downsampling_factor - 1`cells (of downsampled size) beyond your original map (if you use `ceil`s in `updateCostmapSize`) or at most `downsampling_factor - 2`cells (of downsampled size) below your original map (if you use `floor`s in `updateCostmapSize`). See example below.

>  It would be fine from the rounding if around the edges we had 1 cell (at most) extra because of rounding error

This really depends on the size of the original map and the downsampling factor you want to use. Let's picture two extreme examples, that recap what we've learned:

1. Downsample a 10x10 costmap of cells of 0.1m by a sampling factor of 9 -> if we use `ceil`s in `updateCostmapSize`, we would then create a 2x2 downsampled map with cells of 0.9m. This map is 8 cells bigger than the original costmap. Also, in this case we **must** use `min`s when assigning costs to the downsampled costmap to don't retrieve costs fromt the original map that go out-of-bounds (undefined behaviour)
2. Downsample a 10x10 costmap of cells of 0.1m by a sampling by a sampling factor of 6 -> if we use `floor`s in `updateCostmapSize`, we would create a 1x1 downsampled map with cell of 0.6m. This map is 4 cells smaller than the original costmap. We **don't** need to use `min`s in this case, since we won't go out-of-bounds for the original costmap.

Of course, in practice these extreme examples will be rare or non-existent. But they show what are our two options regarding rounding errors for the downsampled costmap size. So to recap our two options are:

1. Have a potentially bigger map and use `min`s when assigning costs to the downsampled costmap
2. Have a potentially smaller map and don't use `min`s when assigning costs to the downsampled costmap

Reply:
I still don't understand why you would ever be out of bounds if you're correctly creating the arrays to your specific size and then using the sizes you know to access them later. Costmaps, smac planner before this, and navfn do not have this issue. I don't think you've explained why this is necessary - rather than patching it, it should be completely removed. Where are these old dimensions even coming from? The new costmap should be a new one of the new size so there should be no knowledge of the prior sizes in use.

If your costmap isn't an exact round integer multiple of the factor, then that barrier cell comprised of only 1 row (or a few rows) should still be populated with the maximum of the data that is in its bounding box (that 1 or few incomplete rows). Either way, that doesn't explain how its possible that a costmap of size `145` or something would let search happen in cells `500+`, when it should know absolutely nothing about those numbers to even think of looking there. 

Are you doing `_costmap[1000 * 1000]` in the code or just an example? That should not be anywhere in the code, there should not be hard coded numbers not based on the input data. 

Your examples make no sense to me. You should never have a **downsampled** map that is ever bigger than the original costmap, that's illogical. If you have a 10x10 0.1m map, any downsampling will always be smaller in size and by a deterministic amount. If you downsample by a non-even number, then the rounding will round up exactly 0 or 1 columns and rows when you start in a corner. That potentialy 1 column or row "border" is still comprised of some known data columns or rows that can be used to populate real valued data.  Either way, that new map will never be a larger size than 10 and it would never cause an overflow accessing if you know the ceil-ed height and width of the downsampled costmap, which you exactly know because you made it

Reply:
Sorry for the misunderstandings. 

> that doesn't explain how its possible that a costmap of size 145 or something would let search happen in cells 500+

Specifically, in [this line](https://github.com/carlosluisg/navigation2/blob/costmap_downsampler/smac_planner/include/smac_planner/costmap_downsampler.hpp#L166) we assign costs to the **downsampled** map's cells by searching the corresponding cells of the **original, full-resolution** map. That's why we end up exploring those cells on the edges of the original costmap, as part of the construction of the downsampled map, it has nothing to do with the planner or anything else.

My wrong assumption in this whole thing was that the size of the downsampled costmap could be bigger than the original map (due to rounding with ceils intead of floors when computing the size in cells, [here](https://github.com/carlosluisg/navigation2/blob/costmap_downsampler/smac_planner/include/smac_planner/costmap_downsampler.hpp#L131]) which could result in retrieving the costs for out-of-bounds cells of the original costmap (resulting in undefined behaviour). Just to clarify, in my current implementation this undefined behaviour will never happen because I used min's [here](https://github.com/carlosluisg/navigation2/blob/costmap_downsampler/smac_planner/include/smac_planner/costmap_downsampler.hpp#L131).

I know understand I shouldn't do this and instead aim for the downsampled cosmtap to be the largest map that fits within the original size, i.e., use floors in the rounding [here](https://github.com/carlosluisg/navigation2/blob/costmap_downsampler/smac_planner/include/smac_planner/costmap_downsampler.hpp#L131), this ensures that when executing [this line](https://github.com/carlosluisg/navigation2/blob/costmap_downsampler/smac_planner/include/smac_planner/costmap_downsampler.hpp#L166) we will never explore out-of-bounds cells of the original costmap, and allow us to remove the mins [here](https://github.com/carlosluisg/navigation2/blob/costmap_downsampler/smac_planner/include/smac_planner/costmap_downsampler.hpp#L165). 

> Are you doing _costmap[1000 * 1000] in the code or just an example?

Just an example for me to verify what I read online about accessing out-of-bounds indeces in C++ arrays, nothing more.

> Your examples make no sense to me. You should never have a downsampled map that is ever bigger than the original costmap, that's illogical. If you have a 10x10 0.1m map, any downsampling will always be smaller in size and by a deterministic amount. If you downsample by a non-even number, then the rounding will round up exactly 0 or 1 columns and rows when you start in a corner. That potentialy 1 column or row "border" is still comprised of some known data columns or rows that can be used to populate real valued data. Either way, that new map will never be a larger size than 10 and it would never cause an overflow accessing if you know the ceil-ed height and width of the downsampled costmap, which you exactly know because you made it

Got it. As you pointed in previous comments, I can remove the min's by doing the correct rounding when assigning the size in cells of the downsampled costmap, which would be doing floors in [these lines](https://github.com/carlosluisg/navigation2/blob/costmap_downsampler/smac_planner/include/smac_planner/costmap_downsampler.hpp#L131-L132). That's my current few-lines diff to have a more optimized code.

Let me know if this answers your concerns. Thanks for the review and sorry again for the misunderstandings.

Reply:
> That's why we end up exploring those cells on the edges of the original costmap, as part of the construction of the downsampled map, it has nothing to do with the planner or anything else.

You know the size of the map in X and Y, you can trivially do boundary checking to skip any indices outside the valid bounds. See the getIndex methods in costmap 2d, there are boundary checking versions of it. In fact, I even wonder when you have a window you're looking at, lets say 3x3 for downsampling, if a row (1x3) is invalid, should we call the value at that cell the max at the 3x2 valid, or call the entire thing LETHAL because it goes off the map. I'm thinking maybe LETHAL.

Reply:
It sounds to me, and please correct me if I'm wrong, that the cause of these bad checks / weird searches are because you're not checking boundaries to find mx and my and then getting its cost here: https://github.com/carlosluisg/navigation2/blob/costmap_downsampler/smac_planner/include/smac_planner/costmap_downsampler.hpp#L166 is that correct? If you did boundary checking, would that then fix any out of bound issues in searching and make sure that the costmap knows nothing about its prior full size?

Reply:
In the current state of the PR there's no real risk of going out of bounds in the line you mentioned. The checks are implicitly made via the `min` calls [here](https://github.com/carlosluisg/navigation2/blob/costmap_downsampler/smac_planner/include/smac_planner/costmap_downsampler.hpp#L164-L165). We are **not** doing any out-of-bound searches on the original costmap thanks to those checks, you can see we limit mx and my to be within the bounds. 

The current "problem" or discussion is how to remove those boundary checks to optimize a bit more the downsampler (you suggested the optimization after my latest commit). The answer is pretty simple, we just have to change to `floor`in [these lines](https://github.com/carlosluisg/navigation2/blob/costmap_downsampler/smac_planner/include/smac_planner/costmap_downsampler.hpp#L131-L132). With that we make sure that always `mx < size_x` and `my < size_y` without using the `min` function. The whole discussion before about going out of bounds was just me testing what happened if I just removed the `min`calls when calculating the `mx`and `my`variables. It doesn't reflect a current risk when using the downsampler.

The downsampled costmap, after creation, doesn't need to know anything about its prior full size. We just use it as a regular costmap within A*.

Reply:
@SteveMacenski any thoughts / concerns with the above comment? If not I'll send a commit with the changes I mentioned.

Reply:
Oh sorry, I don't know how I missed that comment (damn you github!). 

I suppose a good summary is in order - we've covered a bit of ground and I'm a little confused where we stand right now. You mentioned there's some exploring out of bounds - where is that happening? In the planning or in this set costs thing because we're not doing boundry checks? If here, why not add boundary checks so that we know deterministically when its out of bounds to skip non-existent elements of a partial downsampled-cell? I think that's where we're at so that we're not wrapping around.

The floor instead of ceiling comment would then just clip the partial information cells, correct? Is that really what we want to do, or for a partial cell (where there's only 1x3 or 2x3 of the 3x3 of information possible for the window) do we want to fill it in with information based on what is known? I don't know the answer to that question, I could see both.  

What I'm concerned about is reports above about not correctly doing boundary checks so that either:
- We're searching / expanding / exploring cells that are totally bogus (the 500+ numbers for a 200-something downsampled map)
- We're populating downsampled costmap with bogus values from wrap around issues (we're wrapping around for cells on the other side of the map because not boundary checking)

Reply:
It should be easy enough knowing the total costmap size in X and Y to check if the cell is a multiple of it to know if we wrap around and skip those cells. We do this all the time in the costmap_2d codebase to not wrap around on the linear char pointers

Reply:
In the current state of the PR there's no exploring out of bounds, we are checking that the cells explored when assigning costs to the downsampled map are within the bounds of the original map, so there's no way to access bogus values. At planning time we use the downsampled map and there's no risk of going out of bounds on that map either.

The reports you mentioned about not doing boundary checks were just some experiments I conducted at the time, it doesn't reflect what truly happens in the code (it seems they brought more confusion than anything else ^.^"). More specifically, I removed the clipping [here](https://github.com/carlosluisg/navigation2/blob/costmap_downsampler/smac_planner/include/smac_planner/costmap_downsampler.hpp#L164-L165) and ran the code. This came as a suggestion from you at the time to try to remove the clipping as an optimization of the code.

> The floor instead of ceiling comment would then just clip the partial information cells, correct? Is that really what we want to do, or for a partial cell (where there's only 1x3 or 2x3 of the 3x3 of information possible for the window) do we want to fill it in with information based on what is known? I don't know the answer to that question, I could see both.

Currently we are doing the second option: filling a partial cell with the information known for that cell. I'm okay keeping it this way.

> It should be easy enough knowing the total costmap size in X and Y to check if the cell is a multiple of it to know if we wrap around and skip those cells. We do this all the time in the costmap_2d codebase to not wrap around on the linear char pointers

Isn't that equivalent to the clipping already done when calculating the cells to explore in the costmap ([here](https://github.com/carlosluisg/navigation2/blob/costmap_downsampler/smac_planner/include/smac_planner/costmap_downsampler.hpp#L164-L165))? For example, I found [this piece of code](https://github.com/carlosluisg/navigation2/blob/costmap_downsampler/nav2_costmap_2d/src/costmap_2d.cpp#L248) in the costmap_2d codebase which is essentially doing the same clipping, only that it does it with `if` statements rather than `min` calls as I do. Is there any optimization to doing it that way?

Given that in the current code we are not actually exploring any out-of-bounds cells for any of the costmaps (neither the original one nor the downsampled one), is there something else I'm missing? I'm just trying to understand if there are additional concerns related to the way we populate costs in the downsampled map

Reply:
I think this is fine now if that's all accurate. I thought there was something still weird happening here. Sorry. This was prolonged long enough that I lost memory of the issues.